### PR TITLE
cgroup: add Snap/Cgroup monitoring capabilities

### PR DESCRIPTION
**Pull request by sergio-costas**
This MR adds a Snap/Cgroup monitor that allows other parts of the code to wait until all the instances of a Snap have ended and it can be safely refreshed. It is the first step for a bigger MR that will allow to automagically refresh a snap after the system has notified the user that there is a refresh available and that they must close it.

This MR and a future one overrides https://github.com/snapcore/snapd/pull/12155 . The changes from that MR will be proposed in a step-by-step manner to simplify the revision process.

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

### CODE REVIEW

File: sandbox/cgroup/scanning.go [hunk 1]
```diff
@@ -167,6 +159,45 @@ func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
...
+	paths, err := InstancePathsOfSnap(snapInstanceName, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, path := range paths {
+		pids, err := pidsInFile(path)
+		if err != nil {
+			continue
+		}
```
Comment (mardy) [inline sandbox/cgroup/scanning.go:195]: I wonder if we should remove these lines and just add an empty pid list to the `pidsByTag` map; after all, a tag was found, so I think this can only happen if a pid has just exited.

My worry is that if right after InstancePathsOfSnap the only pid present in the only cgroup of the snap exited, and at the same time another pid was started in another cgroup of the same snap, we would return an empty `pidsByTag` map here. Though, maybe, this is not a problem -- it's still not clear to me how the result from this function is used.

File: sandbox/cgroup/monitor.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	watcher: nil,
+	channel: make(chan appData),
+	watched: make(map[string][]*appData),
+}
+
+func monitorMainLoop() {
+	for {
+		select {
+		case event := <-currentCGroupMonitor.watcher.Event:
+			if event.Mask&inotify.InDelete != 0 {
```
Comment (mardy) [inline sandbox/cgroup/monitor.go:56]: I would vote for moving the bodies of these case blocks into separate functions, as the level of indentation here is already quite high.

File: sandbox/cgroup/scanning.go (hunk 1 above)
Comment (sergio-costas) [reply sandbox/cgroup/scanning.go:195]: The problem is... what happens if we find a tag, but between `InstancePathsOfSnap()` and the loop, that snap exits... In that case, when it arrives to the loop, that file doesn't exist.

Anyway, this `PidsOfSnap()` function did exist before; what I did is to extract part of it (to the `InstancePathsOfSnap()` function) to reuse it and don't have the same code duplicated in two places.

File: sandbox/cgroup/monitor.go (hunk 2 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go:112]: Done.

File: sandbox/cgroup/monitor.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	} else {
+		delete(currentCGroupMonitor.watched, basePath)
+		currentCGroupMonitor.watcher.RemoveWatch(basePath)
+	}
+}
+
+func addFiles(newApp *appData) {
+	if newApp.npaths == 0 {
+		newApp.channel <- newApp.name
+	} else {
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor.go:82]: I'd add a `return` statement in the if and then remove this `else`. Then you can remove one level of indentation.

File: sandbox/cgroup/monitor.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	channel chan appData
+}
+
+var currentCGroupMonitor = CGroupMonitor{
+	watcher: nil,
+	channel: make(chan appData),
+	watched: make(map[string][]*appData),
+}
+
+func deletedFile(filename string) {
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor.go:52]: Same for this function, `deletedFile` is a bit too generic and doesn't explain the purpose of the function, a more descriptive name and maybe a doc-string would be nice

File: sandbox/cgroup/monitor.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+import (
+	"errors"
+	"os"
+	"path"
+
+	"github.com/snapcore/snapd/sandbox/cgroup/inotify"
+)
+
+type appData struct {
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor.go:30]: I'm unsure of this name, what about

```suggestion
type appMonitorData struct {
```

Also this structure could use a short doc-string explaining it's purpose

File: sandbox/cgroup/monitor.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+	name        string
+	cgroupPaths []string
+	npaths      int
+	channel     chan string
+}
+
+// This class allows to monitor a CGroup and detect when all the running
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor.go:37]:
```suggestion
// CGroupMonitor allows to monitor a CGroup and detect when all the running
```

File: sandbox/cgroup/monitor.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+// instances have been closed
+
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor.go:39]: remove this rogue newline

File: sandbox/cgroup/monitor.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+		for _, folder := range app.cgroupPaths {
+			if folder == filename {
+				app.npaths--
+			}
+		}
+		if app.npaths == 0 {
+			// all the folders have disappeared, so notify that this app has no more instances running
+			app.channel <- app.name
+		} else {
+			if app.npaths > 0 {
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor.go:66]: You can remove this if statement, because when you are inside the else, you already know that npaths is not zero, which means it must be above 0.

File: sandbox/cgroup/monitor.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+func addFiles(newApp *appData) {
+	if newApp.npaths == 0 {
+		newApp.channel <- newApp.name
+	} else {
+		addedPaths := false
+		for _, fullPath := range newApp.cgroupPaths {
+			basePath := path.Dir(fullPath) // Monitor the path containing this folder
+			_, exists := currentCGroupMonitor.watched[basePath]
+			if !exists {
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor.go:87]: reverse this check and we can remove one level of indentation
```suggestion
			if exists {
			        continue
			}
```

File: sandbox/cgroup/monitor.go [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+				deletedFile(event.Name)
+			}
+		case newApp := <-currentCGroupMonitor.channel:
+			addFiles(&newApp)
+		}
+	}
+}
+
+// Launches the main loop and returns the CGroup singleton
+
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Done.

File: sandbox/cgroup/monitor.go (hunk 5 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Done.

File: sandbox/cgroup/monitor.go (hunk 7 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Done

File: sandbox/cgroup/monitor.go (hunk 9 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Done.

File: sandbox/cgroup/inotify/inotify_linux.go [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+		return errors.New("inotify instance already closed")
+	}
+
+	watchEntry, found := w.watches[path]
+	if found {
+		watchEntry.flags |= flags
+		flags |= syscall.IN_MASK_ADD
+	}
+
+	w.mu.Lock() // synchronize with readEvents goroutine
```
Comment (mardy) [inline sandbox/cgroup/inotify/inotify_linux.go:91]: It's not very clear what this lock is protecting. Looking at other parts of the code, it seems that we care abou the `w.paths` map only. In other words, can we call `syscall.InotifyAddWatch()` outside of the lock?

It seems to me that it would be fine, because then `readEvents()` would just fail on the path lookup and ignore the event.

File: sandbox/cgroup/inotify/inotify_linux.go [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+	return nil
+}
+
+// readEvents reads from the inotify file descriptor, converts the
+// received events into Event objects and sends them via the Event channel
+func (w *Watcher) readEvents() {
+	var buf [syscall.SizeofInotifyEvent * 4096]byte
+
+	for {
+		n, err := syscall.Read(w.fd, buf[:])
```
Comment (mardy) [inline sandbox/cgroup/inotify/inotify_linux.go:144]: Can you please add a comment here, specifying whether this Read() is blocking? (I guess it is)

File: sandbox/cgroup/inotify/inotify_linux.go [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+		}
+		if n < syscall.SizeofInotifyEvent {
+			w.Error <- errors.New("inotify: short read in readEvents()")
+			continue
+		}
+
+		var offset uint32
+		// We don't know how many events we just read into the buffer
+		// While the offset points to at least one whole event...
+		for offset <= uint32(n-syscall.SizeofInotifyEvent) {
```
Comment (mardy) [inline sandbox/cgroup/inotify/inotify_linux.go:176]: Nitpick: I find this easier to read:
```suggestion
		for offset + syscall.SizeofInotifyEvent <= n {
```
(add the cast if/where needed)

File: sandbox/cgroup/monitor.go (hunk 4 above)
Comment (mardy) [reply sandbox/cgroup/monitor.go:56]: Maybe `onFileAdded` and `onFileDeleted`?

**Review by mardy — COMMENTED** [review]
I realized I commented on the inotify code, which you only imported. :-)

So I guess you can ignore these comments, or maybe address them in a separate commit, so we clearly distinguish what changes where made by us.

File: sandbox/cgroup/inotify/inotify_linux.go [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+			w.mu.Lock()
+			name, ok := w.paths[int(raw.Wd)]
+			w.mu.Unlock()
+			if ok {
+				event.Name = name
+				if nameLen > 0 {
+					// Point "bytes" at the first byte of the filename
+					bytes := (*[syscall.PathMax]byte)(unsafe.Pointer(&buf[offset+syscall.SizeofInotifyEvent]))
+					// The filename is padded with NUL bytes. TrimRight() gets rid of those.
+					event.Name += "/" + strings.TrimRight(string(bytes[0:nameLen]), "\000")
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply sandbox/cgroup/inotify/inotify_linux.go:199]: Added a better comment.

File: sandbox/cgroup/inotify/inotify_linux.go (hunk 13 above)
Comment (sergio-costas) [reply sandbox/cgroup/inotify/inotify_linux.go:178]: Honestly, I prefer to avoid touching that part, because it is a low-level code... You know... "if it works"...

File: sandbox/cgroup/monitor.go [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+var currentCGroupMonitor = CGroupMonitor{
+	watcher: nil,
+	channel: make(chan appMonitorData),
+	watched: make(map[string][]*appMonitorData),
+}
+
+func onFilesDeleted(filename string) {
+	basePath := path.Dir(filename)
+	entry := currentCGroupMonitor.watched[basePath]
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor.go:58]: Something more descriptive than `entry`
```suggestion
	appWatchers := currentCGroupMonitor.watched[basePath]
```

File: sandbox/cgroup/monitor.go (hunk 3 above)
Comment (Meulengracht) [reply sandbox/cgroup/monitor.go:84]: still would like this to be addressed

File: sandbox/cgroup/monitor.go [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+			}
+		case newApp := <-currentCGroupMonitor.channel:
+			addFiles(&newApp)
+		}
+	}
+}
+
+// Launches the main loop and returns the CGroup singleton
+
+func GetDefaultCGroupMonitor() *CGroupMonitor {
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply sandbox/cgroup/monitor.go:125]: also this

File: sandbox/cgroup/monitor_test.go [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+var _ = Suite(&monitorSuite{})
+
+func (s *monitorSuite) SetUpTest(c *C) {
+	s.BaseTest.SetUpTest(c)
+
+	s.monitor = cgroup.GetDefaultCGroupMonitor()
+	c.Assert(s.monitor, NotNil)
+}
+
+func (s *monitorSuite) TestMonitorSnap1(c *C) {
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor_test.go:48]: Please use a more descriptive name than numbers in the tests

File: sandbox/cgroup/monitor_test.go [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+		}
+	}
+
+	os.Remove(tmpfile.Name())
+	event := <-channel
+	c.Assert(event, Equals, "test1")
+	c.Assert(s.monitor.NumberOfWaitingMonitors(), Equals, 0)
+}
+
+func (s *monitorSuite) TestMonitorSnap2(c *C) {
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor_test.go:75]: same here

File: sandbox/cgroup/monitor_test.go [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+		}
+	}
+	os.Remove(tmpfile2.Name())
+
+	event := <-channel
+	c.Assert(event, Equals, "test2")
+	c.Assert(s.monitor.NumberOfWaitingMonitors(), Equals, 0)
+}
+
+func (s *monitorSuite) TestMonitorSnap3(c *C) {
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor_test.go:113]: same

File: sandbox/cgroup/scanning.go [hunk 20]
```diff
@@ -167,6 +159,45 @@ func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
...
+//
+// Importantly, if the per-snap lock is held while computing the set, then the
+// following guarantee is true: if a security tag is not among the results then
+// no such tag can come into existence while the lock is held.
+//
+// This can be used to classify the activity of a given snap into activity
+// classes, based on the nature of the security tags encountered.
+func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
+	// pidsByTag maps security tag to a list of pids.
+	pidsByTag := make(map[string][]int)
```
Comment (Meulengracht) [inline sandbox/cgroup/scanning.go:184]: move the declaration of `pidsByTag` down before the for loop, where it's used.

**Review by Meulengracht — COMMENTED** [review]
Thank you for all the changes, I think you missed a few, and added some comments

File: sandbox/cgroup/monitor.go (hunk 15 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Done

File: sandbox/cgroup/monitor_test.go (hunk 19 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor_test.go]: Done too.

File: sandbox/cgroup/monitor_test.go (hunk 18 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor_test.go]: And done too.

File: sandbox/cgroup/scanning.go (hunk 20 above)
Comment (sergio-costas) [reply sandbox/cgroup/scanning.go]: Done.

**Comment by sergio-costas** [comment]
@Meulengracht Done. About those two, I didn't miss them... I don't know why did they still appear to you. (maybe I fix them but forgot to do a push...?)

**Comment by sergio-costas** [comment]
@mardy All done.

**Comment by Meulengracht** [comment]
> @Meulengracht Done. About those two, I didn't miss them... I don't know why did they still appear to you. (maybe I fix them but forgot to do a push...?)

They still appear here in this PR because they are not changed here, if they appear different to you, you may have messed up your branch locally. Take a look at your branch online, it appears just as here (https://github.com/sergio-costas/snapd/blob/DT-576-snapd-refresh-awareness-ux-implementation_add-file-monitoring-support/sandbox/cgroup/monitor.go#L84)

**Comment by sergio-costas** [comment]
@Meulengracht This is what I see in your link, which is the same that I see in my local code...

imagen

You asked me to remove an "if newApp.npaths < 0" that was unneeded, and it isn't there... Or was there another change to do that I missed?

**Comment by Meulengracht** [comment]
> @Meulengracht This is what I see in your link, which is the same that I see in my local code...
> 
> imagen
> 
> You asked me to remove an "if newApp.npaths < 0" that was unneeded, and it isn't there... Or was there another change to do that I missed?

Ah, I think we are talking past each other :-) I was talking about adding a return inside the `if` statement and then removing the 
else, since it isn't needed.

File: sandbox/cgroup/monitor.go [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+	if len(newList) != 0 {
+		currentCGroupMonitor.watched[basePath] = newList
+	} else {
+		delete(currentCGroupMonitor.watched, basePath)
+		currentCGroupMonitor.watcher.RemoveWatch(basePath)
+	}
+}
+
+func onFilesAdded(newApp *appMonitorData) {
+	if newApp.npaths == 0 {
```
Comment (Meulengracht) [inline sandbox/cgroup/monitor.go:82]:
```suggestion
	if newApp.npaths == 0 {
		newApp.channel <- newApp.name
	        return
	}
	.....
```

Comment (sergio-costas) [reply sandbox/cgroup/monitor.go:82]: Oh, ok. Now it makes sense. Done.

File: sandbox/cgroup/monitor.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Done.

File: sandbox/cgroup/monitor.go [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+
+import (
+	"errors"
+	"os"
+	"path"
+
+	"github.com/snapcore/snapd/sandbox/cgroup/inotify"
+)
+
+// appMonitorData contains all the data to monitor an specific Snap:
```
Comment (mardy) [inline sandbox/cgroup/monitor.go:30]: nitpick
```suggestion
// appMonitorData contains all the data to monitor a specific Snap:
```

File: sandbox/cgroup/monitor.go [hunk 23]
```diff
@@ -0,0 +1,171 @@
...
+				onFilesDeleted(event.Name)
+			}
+		case newApp := <-currentCGroupMonitor.channel:
+			onFilesAdded(&newApp)
+		}
+	}
+}
+
+// DefaultCGroupMonitor launches the main loop and returns the CGroup singleton
+func DefaultCGroupMonitor() *CGroupMonitor {
```
Comment (mardy) [inline sandbox/cgroup/monitor.go:125]: The "main loop" is an implementation detail of this class, I'd replace it in the comments with "monitoring loop".

Also, the `DefaultCGroupMonitor()` name suggests that we'll have other kings of cgroup monitors; maybe just call it `Monitor` instead?
Other possibilities (@pedronis might have an opinion):
- GetMonitor
- EnsureMonitor
- MonitorSingleton
- MonitorInstance

File: sandbox/cgroup/monitor.go [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	return &currentCGroupMonitor
+}
+
+// MonitorSnap is the method to call to monitor the running instances of an specific Snap.
+// It receives the name of the snap to monitor (for example, "firefox" or "steam")
+// and a channel. The caller can wait on the channel, and when all the instances of
+// the specific snap have ended, the name of the snap will be sent through the channel.
+// This allows to use the same channel to monitor several snaps
+func (this CGroupMonitor) MonitorSnap(snapName string, channel chan string) {
```
Comment (mardy) [inline sandbox/cgroup/monitor.go:142]: We usually don't use `this` as variable name (though it's fine for me if you keep it as it is), but you definitely need to change all the methods to be 
```suggestion
func (this *CGroupMonitor) MonitorSnap(snapName string, channel chan string) {
```
You can check the assembly at https://godbolt.org/z/n1914d8dr with and without the `*` to see what difference it makes. :-)

File: sandbox/cgroup/monitor.go [hunk 25]
```diff
@@ -0,0 +1,171 @@
...
+	data := appMonitorData{
+		name:        snapName,
+		cgroupPaths: paths,
+		channel:     channel,
+		npaths:      len(paths),
+	}
+	this.channel <- data
+}
+
+// MonitorFiles is currently used for testing. It allows to monitor a group of files/folders
```
Comment (mardy) [inline sandbox/cgroup/monitor.go:153]: Could this be moved to `export_test.go` then?

File: sandbox/cgroup/monitor.go [hunk 26]
```diff
@@ -0,0 +1,171 @@
...
+	data := appMonitorData{
+		name:        name,
+		cgroupPaths: folders,
+		channel:     channel,
+		npaths:      len(folders),
+	}
+	this.channel <- data
+}
+
+// NumberOfWaitingMonitors is currently used for testing. It returns the number of folders being
```
Comment (mardy) [inline sandbox/cgroup/monitor.go:165]: same

File: sandbox/cgroup/monitor.go (hunk 23 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Good point... I think that MonitorSingleton better reflects what it is.

Done.

File: sandbox/cgroup/monitor.go (hunk 24 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: I see... If I understand it right, without the * it passes a copy of the structure/object, while with the * it passes a reference.

Done.

File: sandbox/cgroup/monitor.go (hunk 25 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: It makes sense, yes.

Done.

File: sandbox/cgroup/monitor.go [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+// the running instances of each one have been closed, and notify
+// them separately. It should be considered a singleton, and
+// obtained using MonitorSingleton().
+type CGroupMonitor struct {
+	watched map[string][]*appMonitorData
+	watcher *inotify.Watcher
+	channel chan appMonitorData
+}
+
+var currentCGroupMonitor = CGroupMonitor{
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:50]: it's unclear why we want to have a singleton, we try to avoid that

File: sandbox/cgroup/monitor.go [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+	"os"
+	"path"
+
+	"github.com/snapcore/snapd/sandbox/cgroup/inotify"
+)
+
+// appMonitorData contains all the data to monitor a specific Snap:
+// its name, the list of paths to monitor, and the channel to send
+// the notification when all the paths have been deleted.
+type appMonitorData struct {
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:33]: snapAppMonitorState?

File: sandbox/cgroup/monitor.go [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+	name        string
+	cgroupPaths []string
+	npaths      int
+	channel     chan string
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:37]: this needs a name based on purpose

File: sandbox/cgroup/monitor.go [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+// CGroupMonitor allows to monitor several CGroups, detect when all
+// the running instances of each one have been closed, and notify
+// them separately. It should be considered a singleton, and
+// obtained using MonitorSingleton().
+type CGroupMonitor struct {
+	watched map[string][]*appMonitorData
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:45]: s/watched/monitored/ for consistency?

this also needs a doc comment explaining what the key is

File: sandbox/cgroup/monitor.go [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+	watcher *inotify.Watcher
+	channel chan appMonitorData
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:47]: this also need a clearer name

File: sandbox/cgroup/monitor.go [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+var currentCGroupMonitor = CGroupMonitor{
+	watcher: nil,
+	channel: make(chan appMonitorData),
+	watched: make(map[string][]*appMonitorData),
+}
+
+func onFilesDeleted(filename string) {
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:56]: this should be a method on CGroupMonitor

File: sandbox/cgroup/monitor.go [hunk 33]
```diff
@@ -0,0 +1,151 @@
...
+		return
+	}
+	addedPaths := false
+	for _, fullPath := range newApp.cgroupPaths {
+		basePath := path.Dir(fullPath) // Monitor the path containing this folder
+		_, exists := currentCGroupMonitor.watched[basePath]
+		if exists {
+			continue
+		}
+		err := currentCGroupMonitor.watcher.AddWatch(basePath, inotify.InDelete)
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:93]: maybe a comment here like:

// we watch for deletion happening in basePath

File: sandbox/cgroup/monitor.go [hunk 34]
```diff
@@ -0,0 +1,151 @@
...
+func onFilesAdded(newApp *appMonitorData) {
+	if newApp.npaths == 0 {
+		newApp.channel <- newApp.name
+		return
+	}
+	addedPaths := false
+	for _, fullPath := range newApp.cgroupPaths {
+		basePath := path.Dir(fullPath) // Monitor the path containing this folder
+		_, exists := currentCGroupMonitor.watched[basePath]
+		if exists {
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:90]: given that the check is based only on basePath shouldn't we still check if we need to add things to watched? is there some implicit assumption about the input that is not spelled out?

File: sandbox/cgroup/monitor.go [hunk 35]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	return &currentCGroupMonitor
+}
+
+// MonitorSnap is the method to call to monitor the running instances of an specific Snap.
+// It receives the name of the snap to monitor (for example, "firefox" or "steam")
+// and a channel. The caller can wait on the channel, and when all the instances of
+// the specific snap have ended, the name of the snap will be sent through the channel.
+// This allows to use the same channel to monitor several snaps
+func (this *CGroupMonitor) MonitorSnap(snapName string, channel chan string) {
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:142]: we follow golang naming conventions, this should be m for monitor

File: sandbox/cgroup/monitor.go [hunk 36]
```diff
@@ -0,0 +1,151 @@
...
+	paths, _ := InstancePathsOfSnap(snapName, InstancePathsFlagsOnlyPaths)
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:143]: we almost never ignore errors, and when we do we need to explain why and still log them

File: sandbox/cgroup/monitor.go [hunk 37]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/snapcore/snapd/sandbox/cgroup/inotify"
+)
+
+// appMonitorData contains all the data to monitor a specific Snap:
+// its name, the list of paths to monitor, and the channel to send
+// the notification when all the paths have been deleted.
+type appMonitorData struct {
+	name        string
+	cgroupPaths []string
+	npaths      int
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:36]: maybe this two should be a cgroupPaths map[string]bool and we remove the entries when the we get the deletion notifications?

File: sandbox/cgroup/monitor.go [hunk 38]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	if len(newList) != 0 {
+		currentCGroupMonitor.watched[basePath] = newList
+	} else {
+		delete(currentCGroupMonitor.watched, basePath)
+		currentCGroupMonitor.watcher.RemoveWatch(basePath)
+	}
+}
+
+func onFilesAdded(newApp *appMonitorData) {
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:81]: this should be a method on CGroupMonitor

File: sandbox/cgroup/monitor.go [hunk 39]
```diff
@@ -0,0 +1,151 @@
...
+		currentCGroupMonitor.watched[basePath] = append(currentCGroupMonitor.watched[basePath], newApp)
+		addedPaths = true
+	}
+	if !addedPaths {
+		// if the files/folders to monitor don't exist, send the notification now
+		newApp.channel <- newApp.name
+	}
+}
+
+func monitorMainLoop() {
```
Comment (pedronis) [inline sandbox/cgroup/monitor.go:111]: this should be a method on CGroupMonitor

File: sandbox/cgroup/monitor.go (hunk 35 above)
Comment (pedronis) [inline sandbox/cgroup/monitor.go:142]: this in itself doesn't seem to be tested

File: sandbox/cgroup/scanning.go [hunk 40]
```diff
@@ -67,33 +68,25 @@ func securityTagFromCgroupPath(path string) naming.SecurityTag {
...
-// security tag. The result may be immediately stale as processes fork and
-// exit.
-//
-// Importantly, if the per-snap lock is held while computing the set, then the
-// following guarantee is true: if a security tag is not among the results then
-// no such tag can come into existence while the lock is held.
+type InstancePathsFlags uint32
+
+const (
+	InstancePathsFlagsOnlyPaths InstancePathsFlags = 1 << iota
```
Comment (pedronis) [inline sandbox/cgroup/scanning.go:74]: we tend to use *Options structs with bool fields for this kind of things

File: sandbox/cgroup/scanning.go [hunk 41]
```diff
@@ -111,7 +104,7 @@ func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
 	// tag matches the snap we are interested in, we harvest the snapshot of
 	// PIDs that belong to the cgroup and put them into a bucket associated
 	// with the security tag.
-	walkFunc := func(path string, fileInfo os.FileInfo, err error) error {
+	walkFunc := func(basepath string, fileInfo os.FileInfo, err error) error {
```
Comment (pedronis) [inline sandbox/cgroup/scanning.go:107]: why this change?

File: sandbox/cgroup/scanning.go [hunk 42]
```diff
@@ -167,6 +159,46 @@ func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
...
+// exit.
+//
+// Importantly, if the per-snap lock is held while computing the set, then the
+// following guarantee is true: if a security tag is not among the results then
+// no such tag can come into existence while the lock is held.
+//
+// This can be used to classify the activity of a given snap into activity
+// classes, based on the nature of the security tags encountered.
+func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
+
```
Comment (pedronis) [inline sandbox/cgroup/scanning.go:183]: this seems a spurious blank line

**Review by pedronis — COMMENTED** [review]
did a pass, bunch of style comments and some questions, we value consistency and clarity quite a bit as snapd is a fairly large codebase

File: sandbox/cgroup/scanning.go [hunk 43]
```diff
@@ -67,33 +68,25 @@ func securityTagFromCgroupPath(path string) naming.SecurityTag {
...
+// InstancePathsOfSnap returns the list of active cgroup paths for a given snap
 //
-// This can be used to classify the activity of a given snap into activity
-// classes, based on the nature of the security tags encountered.
-func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
-	// pidsByTag maps security tag to a list of pids.
-	pidsByTag := make(map[string][]int)
+// The return value is a snapshot of the cgroup paths
+
+func InstancePathsOfSnap(snapInstanceName string, flags InstancePathsFlags) ([]string, error) {
```
Comment (pedronis) [inline sandbox/cgroup/scanning.go:81]: it's not clear we need this exported, either way we probably want new unit tests to show what it returns in the scenarios we already test for PidsOfSnap

File: sandbox/cgroup/monitor.go (hunk 27 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go:51]: Removed the singleton.

File: sandbox/cgroup/monitor.go (hunk 28 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Done.

File: sandbox/cgroup/monitor.go (hunk 30 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Done.

File: sandbox/cgroup/monitor.go (hunk 29 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Done

File: sandbox/cgroup/monitor.go (hunk 35 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: I changed the code. Tell me if the current one is enough.

File: sandbox/cgroup/scanning.go (hunk 41 above)
Comment (sergio-costas) [reply sandbox/cgroup/scanning.go:107]: To avoid clash with the "path" import.

File: sandbox/cgroup/scanning.go (hunk 42 above)
Comment (sergio-costas) [reply sandbox/cgroup/scanning.go]: fixed.

File: sandbox/cgroup/scanning.go (hunk 43 above)
Comment (sergio-costas) [reply sandbox/cgroup/scanning.go:81]: I need it exported to be able to use it in the monitor.

File: sandbox/cgroup/monitor.go (hunk 39 above)
Comment (sergio-costas) [reply sandbox/cgroup/monitor.go]: Done, and also onFilesAdded() and onFilesDeleted()

File: sandbox/cgroup/scanning.go (hunk 40 above)
Comment (sergio-costas) [reply sandbox/cgroup/scanning.go]: Done.

File: sandbox/cgroup/scanning.go (hunk 43 above)
Comment (sergio-costas) [reply sandbox/cgroup/scanning.go]: Added tests.

**Comment by sergio-costas** [comment]
@mardy @pedronis @mvo5 @Meulengracht I wrote an alternative patch for the monitor itself, where I fully remove the singleton. It reuses the changes in cgroup/scanning.go, which were the most reviewed by you, and completely rewrite and simplify the cgroup/monitor.go file.

https://github.com/snapcore/snapd/pull/12280

Which one do you prefer?

**Comment by sergio-costas** [comment]
I marked this one as Draft because https://github.com/snapcore/snapd/pull/12280 is a better implementation.

**Comment by sergio-costas** [comment]
Closing this too because all the work is being done in https://github.com/snapcore/snapd/pull/12280
